xrootd/streammanager/streammanager.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package streammanager // import "go-hep.org/x/hep/xrootd/streammanager"
+
+import (
+	"math"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+type ServerResponse struct {
+	Data  []byte
+	Error error
+}
+
+type dataSendChannel chan<- *ServerResponse
+type DataReceiveChannel <-chan *ServerResponse
+
+const streamIDPartSize = math.MaxUint8
+const streamIDPoolSize = streamIDPartSize * streamIDPartSize
+
+// StreamManager manages channels by their ids.
+// Basically, it's a map[StreamID] chan<-*ServerResponse
+// with methods to claim, free and pass data to specific channel by id.
+type StreamManager struct {
+	dataWaiters map[protocol.StreamID]dataSendChannel
+	freeIds     chan protocol.StreamID
+	mutex       sync.Mutex
+}
+
+// New creates new StreamManager
+func New() *StreamManager {
+	sm := StreamManager{
+		make(map[protocol.StreamID]dataSendChannel),
+		make(chan protocol.StreamID, streamIDPoolSize),
+		sync.Mutex{},
+	}
+
+	for firstByte := uint16(0); firstByte < streamIDPartSize; firstByte++ {
+		for secondByte := uint16(0); secondByte < streamIDPartSize; secondByte++ {
+			sm.freeIds <- protocol.StreamID{byte(firstByte), byte(secondByte)}
+		}
+	}
+
+	return &sm
+}
+
+// Claim searches for unclaimed id and returns corresponding channel
+func (sm *StreamManager) Claim() (id protocol.StreamID, channel DataReceiveChannel, err error) {
+	bidirectionalChannel := make(chan *ServerResponse, 1)
```

Comment:
just curious but: why is this channel buffered?

Reply:
I thought about the case when `SendData` will be deadlocked due to unreceived data from channel. But that's highly unlikely now because all reads from this channel will be done in one method of `Client`. Anyway, we can't specify buffer size (if we are not going to use some huge number for that) because server response can be split in many messages and client doesn't know how many responses expect. I'll make it unbuffered.